rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+
+### Defining storage traits
+
+[defining-storage-traits]: #defining-storage-traits
+
+The clear definition of the `Storage` trait allows defining storage traits. A storage trait is a trait that has `Storage` as a supertrait. The STD will come with (at least) two storage traits, `DeepReadStorage` and `DeepClearStorage` both defined in the [`storage.sw`](../files/0013-configurable-and-composable-storage/sway-libs/storage.sw).
+
+The implementations of storage traits will, similar to the implementations of `Storage`, be straightforward and based on recursive calls to the implementations of the same storage traits on the contained storage types.
+
+To get the feeling for these straightforward and essentially short and compact implementations of storage traits, see the implementations of `DeepReadStorage` and `DeepClearStorage` for [`StorageVec`](../files/0013-configurable-and-composable-storage/sway-libs/storage/storage_vec.sw) and [`StoragePair`](../files/0013-configurable-and-composable-storage/user-defined-libs/storage_pair.sw).
+
+By convention, all storage trait will have the suffix `Storage`.
+
+## Internal API and the `__state` intrinsics
+
+[internal-api-and-the-state-intrinsics]: #internal-api-and-the-state-intrinsics
+
+The STD will still provide the low-level storage `read`, `write`, and `clear` functions. However, unlike now when they are actually sometimes needed in contract code, e.g., to store arrays, in the new storage implementation there should never be a need to use them in contracts. Instead, the atomic `StorageBox` and `StorageEncodedBox` should be used. They provide the same low-level functionality while offering safe storage access.
+
+The low-level API will be used only when implementing storage types, and even in those cases not always. Thus, the proposal is to move them to the module named [`internal`](../files/0013-configurable-and-composable-storage/sway-libs/storage/internal.sw) to emphasize that they are, similar to `Storage::internal_` functions, meant to be used only when developing custom storage types.
+
+Similarly, the `__state_` intrinsics should be used only in the implementations of the internal API functions. In addition, the `__state_load_word` intrinsic should be changed to return the information if the requested slot was set or unset.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+The only drawback I can think of is the time and effort needed to implement the proposal and to deal with the [breaking changes](#breaking-changes). However, the impact of _not_ improving the current storage handling will over time surely be higher. It would mean carrying on the issues mentioned in the [Motivation](#motivation) and, worst of all, living with an API that is error-prone by-design.
```

Comment:
Maybe we should have some discussion of the performance implications of this change. Can storage get slower as a result of this, if employed correctly?

Reply:
I say - no :smile: (to the negative implications, not to the discussion :smile:)

Or at least I don't see any negative impact on the performance, compared to the approach we have now. On the contrary:
- Performance was in focus and considered carefully. It resulted, e.g, in the `StorageLayout` enum being introduced.
- Just using the `StorageLayout` information consistently will help `Storage` implementors to provide implementations that optimally pack and store data across the whole hierarchy of the stored data.
- Having the convention for "special behavior" methods should encourage providing optimized versions of methods where high-performance is needed and safety guaranteed by surrounding code.

But it can be that I am missing something, so yes we can surely discuss it further. And of course reviewing the proposal from the performance standpoint is highly encouraged.